src/mbed_tools/project/mbed_program.py

```diff
@@ -91,6 +91,32 @@ def from_existing(cls, dir_path: Path, check_mbed_os: bool = True) -> "MbedProgr
 
         return cls(program, mbed_os)
 
+    @classmethod
+    def from_existing_with_mbed_os(cls, dir_path: Path, mbed_os_path: Path, check_mbed_os: bool = True) -> "MbedProgram":
```

Comment:
We should probably just add logic to `MbedProgram::from_existing` to handle the mbed path if possible, instead of adding this new method.

Something like:

```
    @classmethod
    def from_existing(cls, dir_path: Path, mbed_os_path: Optional[Path] = None, check_mbed_os: bool = True) -> "MbedProgram":
        """Create an MbedProgram from an existing program directory.
        Args:
            dir_path: Directory containing an Mbed program.
            check_mbed_os: If True causes an exception to be raised if the Mbed OS source directory does not
                           exist.
        Raises:
            ProgramNotFound: An existing program was not found in the path.
        """
        if mbed_os_path is None:
            program_root = _find_program_root(dir_path)
            mbed_os_path = program_root / MBED_OS_DIR_NAME
        else:
            program_root = dir_path

        logger.info(f"Found existing Mbed program at path '{program_root}'")
        program = MbedProgramFiles.from_existing(program_root)

        try:
            mbed_os = MbedOS.from_existing(mbed_os_path, check_mbed_os)
        except ValueError as mbed_os_err:
            raise MbedOSNotFound(
                f"Mbed OS was not found due to the following error: {mbed_os_err}"
                "\nYou may need to resolve the mbed-os.lib reference. You can do this by performing a `deploy`."
            )

        return cls(program, mbed_os)
```

Reply:
@rwalton-arm Thanks for the input! I hadn't seen your message and came to the same solution.

just one question: is the `Optional[Path]` mandatory? I did not use it and it seems to work fine.

I've updated the PR.

Reply:
The type checker `mypy` does support "implicit optional", but this is likely to be removed in a future version. Also for readability we should add it.